### DISCUSSION TITLE
router.push() does not get the latest revalidated page

### DISCUSSION THREAD

**Discussion by anasqadrei** [vercel/next.js#40888] [Help]
Consider this page
```javascript
// pages/index.js
import Link from 'next/link'
import { useRouter } from 'next/router'

export async function getStaticProps() {
  return {
    props: {
      time: new Date().getTime(),
    },
  }
}

export default function Home({ time }) {
  const router = useRouter()

  return (
    <>
      <p>
        {time}
      </p>
      <Link href="/">Link</Link>
      <button onClick={() => { router.push('/') }}>
        push
      </button>
      <button onClick={() => { router.replace('/') }}>
        replace
      </button>
      <button onClick={() => { fetch('/api/revalidate') }}>
        revalidate
      </button>
    </>
  )
}
```
and this API
```javascript
// pages/api/revalidate.js
export default async function handler(req, res) {
  try {
    await res.revalidate('/')
    return res.json({ revalidated: true })
  } catch (err) {
    return res.status(500).send('Error revalidating')
  }
}
```
In that example, `time` is statically generated. To update it, you must click `revalidate` to invalidated the cache using [On-demand Revalidation](https://nextjs.org/#on-demand-revalidation) then refresh the page. That works fine and as intended. 

The problem is I want to programatically refresh the page using `router.push()` or `router.replace()` after `revalidate`. It does not work. After clicking `revalidate` then `push` or `replace`, nothing happens. However, `Link` works. Shouldn't `push()` and `Link` have similar behaviours as documented in [router.push](https://nextjs.org/#routerpush)? Is this a bug or am I missing something?

I'm using `next 12.3.1`, `react 18.2.0` and `node 14.19.1`

Why I have this problem? For SEO purposes, my page gets its content from `getStaticProps`. I let the user update the content. Then, I `revalidate()` and `router.push()` to show the updated content to the user. Sadly, it does not work :(

PS. I don't want to use `router.reload()`. It's ugly. I have music playing in one of the components and reload would stop it

**Comment by octet-stream** [comment]
I think I have the same problem with `replace` and `push`.
I also noticed that when I revalidate page for the first time after the fresh production build was created, the first redirect will result in one `GET` (presumably it loads newly updated data for page) and one `HEAD` requests being sent on page. But when I'm trying to revalidate the page again, then redirect - only `HEAD` request is being send. At the same time, navigating via `next/link` always returns updated data.

For more context:
1. The page has dynamic route with following format: `/post/[date]/[name]`
2. The data is updated from `/post/[date]/[name]/edit` page, and revalidation happens in the same API call that updates the data.
3. After the data is updated, I use `router.replace()` to navigate back to the post page. After the first update on fresh production build everything works as expected - the new page is rendered. But after following updates the page is always served with the old data, except when navigating via `next/link` or refreshing the page.

**Reply by anasqadrei** [reply]
Yup, that's a similar behaviour. I issued a bug [here](https://github.com/vercel/next.js/issues/40927) and was solved but hasn't been released yet. The solution is to use `router.push(url, undefined, { unstable_skipClientCache: true })`

**Reply by octet-stream** [reply]
That's great news! But at the meantime I made a workaround using `createContext` and `useState` hook:
1. When the page component mounts, I create state object using `useState`, where the key it the page url (a value from router.asPath) and value is the page data.
2. Then I create a context with the data for current page + a callback that updates the state for specified page.
3. So when you update some data, you'll need to call this callback with the new data for the page you want to update, and then you just redirect to it.

If somebody will come across this discussion before the fix arrives, you can find this workaround here: [Context and state](https://github.com/octet-stream/eri/blob/5b125dc06a90fc6833a7615ee8ec3147e746e723/src/lib/context/PageDataContext.tsx) and [the example of usage](https://github.com/octet-stream/eri/blob/5b125dc06a90fc6833a7615ee8ec3147e746e723/src/pages/post/%5Bdate%5D/%5Bname%5D/edit.tsx#L87). Of course the context is rendered within `_app.tsx`.

**Comment by vnugent** [comment]
I'm having the same issue with v12.3.4.  After calling revalidate, I can open a new tab and see updated props coming from `getStaticProps`.  Props on the current page, however, are stale. 

*Update:

Issue summary:  
- Version 12.3.4
- Revalidate works fine when viewing the page in another tab/browser window
- The main issue is the current page keeps showing stale props unless you do a hard browser refresh or click on a 'Link' and 'a' tag pointing back to the same page.  

After spending nearly 8hrs trying various things I came up this hack

```javascript
router.replace(`/${Date.now()}`)
```

**Reply by salman0ansari** [reply]
Thank you so much, this fixed the issue.
